### ISSUE TITLE
Initial image previews are too slow

### ISSUE THREAD

**Issue by soulofmischief** [webaverse/app#937]
We should either pre-cache generated images or speed up the generation process. Users encountering new items encounter UI jank as the previews can take a long time to load.

**Comment by avaer** [comment]
I see 3 ways to speed this up, not necessarily mutually exclusive.

The first is to use distributed generation/preview servers, which might require load balancing.
The second is to use expensive GPU servers for the preview.
The third is to use a non-browser solution for previews, like a non-browser render process. This would change the fidelity and the trustworthiness of previews though (since it is running different code from the client), and I don't know how much faster it would be, so it's more of a wildcard.

**Comment by avaer** [comment]
https://github.com/webaverse/app/issues/568

**Comment by avaer** [comment]
Currently still slow:

```
a@DESKTOP-804SN1O:/mnt/c/Users/avaer/Documents/GitHub/webaverse.com$ time curl 'https://preview.exokit.org/Qme4BxHNzo44rnG2wkFsr29EhmM7BQFvXwYWiqNr8af9XF.wbn/preview.png'
Warning: Binary output can mess up your terminal. Use "--output -" to tell
Warning: curl to output it to your terminal anyway, or consider "--output
Warning: <FILE>" to save to a file.

real    0m0.500s
user    0m0.031s
sys     0m0.016s
a@DESKTOP-804SN1O:/mnt/c/Users/avaer/Documents/GitHub/webaverse.com$ ^C
a@DESKTOP-804SN1O:/mnt/c/Users/avaer/Documents/GitHub/webaverse.com$ ^C
a@DESKTOP-804SN1O:/mnt/c/Users/avaer/Documents/GitHub/webaverse.com$ ^C
a@DESKTOP-804SN1O:/mnt/c/Users/avaer/Documents/GitHub/webaverse.com$ time curl 'https://preview.exokit.org/Qme4BxHNzo44rnG2wkFsr29EhmM7BQFvXwYWiqNr8af9XF.wbn/preview.png'
Warning: Binary output can mess up your terminal. Use "--output -" to tell
Warning: curl to output it to your terminal anyway, or consider "--output
Warning: <FILE>" to save to a file.

real    0m5.947s
user    0m0.016s
sys     0m0.000s
```

**Comment by avaer** [comment]
This is a sub-issue of general performance slowness of the frontend: https://github.com/webaverse/app/issues/856

**Comment by avaer** [comment]
What did yesterday’s perf logs reveal about the timing fluctuations? Is it a complicated fix?

**Comment by soulofmischief** [comment]
Well for one thing, I found the preview backend never touches the blockchain when pulling from a cached image on s3, so I've isolated the problem to s3 itself, and I'm not versed in s3 optimization at all so I've been doing some reading and trying things.

**Comment by soulofmischief** [comment]
My sleep schedule has been backwards the last two nights but i'm resetting it tonight bc I'd like to pair up with chris on this tomorrow because this is something he's more familiar with.

**Comment by soulofmischief** [comment]
Could it be regional latency?

**Comment by avaer** [comment]
Did you time the S3 request itself?

**Comment by avaer** [comment]
I don't think it's regional, since requests from the same machine have different timings (correct me if I am wrong).

**Comment by avaer** [comment]
- Make the S3 buckets public so direct tests are possible
- Add a CloudFlare route in front

**Comment by shawticus** [comment]
file with public permissions for testing: https://preview-exokit-org.s3-us-west-1.amazonaws.com/Qme4BxHNzo44rnG2wkFsr29EhmM7BQFvXwYWiqNr8af9XF/wbn/png
raw s3 lookup in curl:
dnslookup: 0.000792s
connect: 0.017082s
appconnect: 0.096555s
pretransfer: 0.096581s
starttransfer: 0.124521s
total: 0.152482s

**Comment by shawticus** [comment]
Speed test of request start to download ->

Testing AWS SDK locally with transfer acceleration OFF:
GET OBJECT: 198.413ms
GET OBJECT: 146.006ms
GET OBJECT: 142.43ms
GET OBJECT: 144.7ms

Testing AWS SDK locally with transfer acceleration ON:
GET OBJECT: 153.418ms
GET OBJECT: 180.925ms
GET OBJECT: 500.573ms
GET OBJECT: 179.42ms

This is in line with the cURL request. Transfer Acceleration should dramatically increase speed in other regions, but could be slower (due to being an additional load balancer on our request)

**Comment by shawticus** [comment]
Previews load for me in 40ms in Chrome, the whole site loads in a lot less than 5 seconds. We shouldn't need transfer acceleration to a regionally connected EC2 machine, what would speed things up for the client would be handing a direct bucket URL back so we don't need to pipe the stream, but at least for me in California this isn't slowing me down much.
So we have 2 solutions:
1. Regional EC2 boxes
2. Hand the user back a public s3 URL directly so we don't need to pipe the request and it can take advantage of local CDN
Clearly, #2 is the winner.
![Screenshot from 2021-03-13 22-23-20](https://user-images.githubusercontent.com/18633264/111059518-197d7c80-844b-11eb-982b-b97ed9a69407.png)

**Comment by soulofmischief** [comment]
Actually using the accelerate bucket from the production server itself still introduces about 40-50ms latency.

```
// US West 1
$ curl -w "dnslookup: %{time_namelookup} | connect: %{time_connect} | appconnect: %{time_appconnect} | pretransfer: %{time_pretransfer} | starttransfer: %{time_starttransfer} | total: %{time_total} | size: %{size_download}\n" -so /dev/null https://preview-exokit-org.s3-us-west-1.amazonaws.com/Qme4BxHNzo44rnG2wkFsr29EhmM7BQFvXwYWiqNr8af9XF/wbn/png

dnslookup: 0.002344 | connect: 0.002587 | appconnect: 0.011747 | pretransfer: 0.011780 | starttransfer: 0.023207 | total: 0.023834 | size: 63818
dnslookup: 0.001495 | connect: 0.002623 | appconnect: 0.014160 | pretransfer: 0.014193 | starttransfer: 0.028888 | total: 0.031613 | size: 63818


// Accelerate
$ curl -w "dnslookup: %{time_namelookup} | connect: %{time_connect} | appconnect: %{time_appconnect} | pretransfer: %{time_pretransfer} | starttransfer: %{time_starttransfer} | total: %{time_total} | size: %{size_download}\n" -so /dev/null https://preview-exokit-org.s3-accelerate.amazonaws.com/Qme4BxHNzo44rnG2wkFsr29EhmM7BQFvXwYWiqNr8af9XF/wbn/png

dnslookup: 0.001436 | connect: 0.002188 | appconnect: 0.012765 | pretransfer: 0.012801 | starttransfer: 0.062843 | total: 0.066076 | size: 63818
dnslookup: 0.001420 | connect: 0.002265 | appconnect: 0.013277 | pretransfer: 0.013315 | starttransfer: 0.073364 | total: 0.074903 | size: 63818
```

**Comment by soulofmischief** [comment]
If we hand the URL to the client, we need to first make sure it's valid and that we don't need to generate and store the preview first.

I started doing this with

```
s3.waitFor('objectExists', params, function(error, data) {
  if (error) {
    // Generate preview
  }
  else return url;
});
```
but `objectExists` seems to use `s3.headObject() under the hood, which requires `s3:ListBucket` permissions enabled on the bucket. I'm not sure this is the right approach. Do we want the bucket to be public and indexable?

Alternatively I could just do an internal request to the url and handle 404s. This smells like bad code but `s3:ListBucket` smells worse.

**Comment by avaer** [comment]
How does `s3.waitFor('objectExists',` handle the case where the preview locked up and isn't coming?

**Comment by soulofmischief** [comment]
it's just a wrapper for `s3.headObject()` which gets the head information of the file, which is the only idiomatic way to check for a file existing in the SDK. Handling process lock / request timeout is out of its scope.

Example: https://stackoverflow.com/a/56092118

**Comment by avaer** [comment]
So then it will lock up and there is a bug?

**Comment by soulofmischief** [comment]
Bug where? I haven't even finished implementing this because I wanted an RFC. There are plenty of ways to handle that condition.

**Comment by avaer** [comment]
Yeah I'm thinking ahead to how this handles the very common case of where the image will fail to generate. It's important to decide when that is triggered and how failures are handled. Using the bucket directly removes a lot of freedoms in how to handle that case.

**Comment by soulofmischief** [comment]
Yeah, I can handle that issue so that's not concerning me. What's concerning is whether or not we are being too open with this bucket

**Comment by avaer** [comment]
I think the openness is fine.

Keep in mind when we provide the preview URL we do not have the choice of how we render it and we cannot "fix up" previews client side (e.g. OpenSea, meta previews, etc.)

**Comment by soulofmischief** [comment]
So with that in mind, do you think we should stick to serving the content from the backend?

**Comment by avaer** [comment]
That was the original reasoning behind that design. I don't know about now, but the requirements haven't changed.

**Comment by soulofmischief** [comment]
Shaw's other suggestion was regional EC2 boxes. That seems like a workaround for some kind of larger issue though. Cloudfront should be solving this.

**Comment by avaer** [comment]
Can you do the S3 solution and make sure preview errors are generated into the image?
